src/tools/mesh_tester.cpp

```diff
@@ -250,6 +328,97 @@ std::vector< std::pair<int, int> > naiveIntersectionAlgorithm(
   return retval;
 }
 
+#if defined(AXOM_USE_RAJA)
+template < typename ExecSpace >
+std::vector< std::pair<int, int> > naiveIntersectionAlgorithm(
+  mint::Mesh* surface_mesh,
+  std::vector<int> & degenerate)
+{
+  SLIC_INFO("Running naive intersection algorithm "
+    << " in execution Space: " 
+    << axom::execution_space< ExecSpace >::name());
+
+  // Get allocator
+  int allocatorID = axom::execution_space< ExecSpace >::allocatorID();
+  axom::setDefaultAllocator( allocatorID );
+
+  std::vector< std::pair<int, int> > retval;
+
+  const int ncells = surface_mesh->getNumberOfCells();
+  SLIC_INFO("Checking mesh with a total of "<< ncells<< " cells.");
+
+  Triangle3 * tris = axom::allocate <Triangle3> (ncells);
+
+  // Get each triangle in the mesh and check for degeneracies
+  for (int i = 0; i < ncells ; i++)
+  {
+    tris[i] = getMeshTriangle(i, surface_mesh);
+    if (tris[i].degenerate())
+    {
+      degenerate.push_back(i);
+    }
+  }
+
+  RAJA::RangeSegment row_range(0, ncells);
+  RAJA::RangeSegment col_range(0, ncells);
+
+  using KERNEL_POL = 
+    typename axom::mint::internal::structured_exec< ExecSpace >::loop2d_policy;
+  using REDUCE_POL =
+    typename axom::execution_space< ExecSpace >::reduce_policy;
+  using ATOMIC_POL = 
+    typename axom::execution_space< ExecSpace >::atomic_policy;
+ 
+  RAJA::ReduceSum< REDUCE_POL, int > numIntersect(0);  
+
+  // Compute the number of intersections
+  RAJA::kernel<KERNEL_POL>( RAJA::make_tuple(col_range, row_range),
+    AXOM_LAMBDA(int col, int row) {
+    if (row > col)
+    {
+      if (checkTT (tris[row], tris[col]))
+      {
+        numIntersect += 1;
+      }
+    }
+  });
+
+  // Allocation to hold intersection pairs and counter to know where to store
+  int * intersections =
+    axom::allocate <int> (numIntersect.get() * 2);
+  int * counter = axom::allocate <int> (1);
+  
+  counter[0] = 0;
+
+  // RAJA loop to populate with intersections
+  RAJA::kernel<KERNEL_POL>( RAJA::make_tuple(col_range, row_range),
+    AXOM_LAMBDA(int col, int row) {
+    if (row > col)
+    {
+      if (checkTT (tris[row], tris[col]))
+      {
+        auto idx = RAJA::atomicAdd<ATOMIC_POL>(counter, 2);
```

Comment:
It seems that you could store a `counts` and `offsets` array during the kernel that computes the number of intersections and get rid of the atomic add here. Thoughts?

Reply:
Or rather do `scan` after you calculate the `counts` array to compute the `offsets`....

Reply:
It is not clear that that would help in this case.

For a triangle mesh with ``N`` triangles, that would cost ``2N`` ints.
For most triangle meshes, we'd expect to have very few intersections.

Reply:
It would remove the need for an atomic. 

There is already a kernel launch prior to this to compute the total count, so you might as well fill in the `counts` array and alleviate the need for an atomic. I'd expect it to be faster.

Reply:
> Or rather do a scan after you calculate a counts array and compute the offsets....

@gzagaris I do not think that I am following correctly and am missing something. The `offsets` would give me the index to write one pair of intersecting triangles. If I had multiple intersections for a given triangle, wouldn't I still need an atomic  to increment the offset for intersecting pairs after the first?

Reply:
Since the current version is working, I think we should defer this sort of optimization to a separate pass.

Reply:
> @gzagaris I do not think that I am following correctly and am missing something. The `offsets` would give me the index to write one pair of intersecting triangles. If I had multiple intersections for a given triangle, wouldn't I still need an atomic to increment the offset for intersecting pairs after the first?

@bmhan12 , yes you are correct. You would need an atomic as the code currently is. 

I thought that in your first kernel you were excluding candidate triangles based on a bounding box intersection and that in your second kernel you were processing an abbreviated list of candidate triangles serially within each thread, in which case, you could just keep a integer counter for each thread....